### ISSUE TITLE
Discussion: Module polymophism contracts with `ABC`

### ISSUE THREAD

**Issue by bfcarpio** [hhursev/recipe-scrapers#317]
Alternate title: "OOPism and you, consistency in the face of developer laziness" :rofl: 

> Note: I'm going to talk about types in the following discussion. I want to make it explicit that at this time I don't intend on making the package "type safe" or "type compliant" etc. Ideally, I'd like to have type hints or a stub available so that consumers can see it.

A couple months ago a large change was made to this repository wherein the convention would be to explicitly call the schema scraper functions. Therefore, each child scraper should ideally have every single function defined with either a custom `bs4` implementation, a schema function call, or a generic `return None` etc call. However, since the decorators filling in missing implementations still exists we have scrapers that are "half-defined" so to speak.

Higher up in my priorities, I've noticed that it's somewhat difficult to consume this package. There is confusion about return types (#283) and, from my experience, documentation and type hints don't show up when using IDE features [1] even though we use a typical OOP inheritance pattern. Yes, I acknowledge, there aren't many doc strings available, but during my exploring I've manually added some. You should still be able to check this with `host()`.

From my testing, I think this might have to do with how Python implements OOP and specifically the need for [`ABC`](https://docs.python.org/3/library/abc.html#abc.ABC). When I added dummy doc strings, made `AbstractScraper` inherit `ABC` and add `@abstractmethod` to the relevant functions, and implemented missing functions to a scraper I was able to get the hover documentation. Additionally, I got errors when trying to instantiate a scraper class that didn't implement all the `@abstractmethod` functions.

So why am I creating a discussion and not making a PR with all this already made? First, I'm not sure how much of a *contract* is wanted for these scrapers. Secondly, and most importantly, we'd have to find a way to re-implement the error silencing feature [2]. Since using `ABC` would force scraper implementations to manually define each function, I think we could remove `@Decorators.schema_org_priority` and replace it with some other decorator with that handles the silencing. I'm not the most sure what the best way to changing this implementation would be. I've already checked a couple ways and I don't think it's possible to keep the metaclass and use `ABC` or `ABCMeta`.

I'm happy to write nice sphinx-style docs and implement what's required to use `ABC` meta, but I'd want to know what's a path for doing error handling.

---
 1. I use `neovim` with `coc-python`. I've also tried it with `coc-pyright`. EDIT: doc hover just utilizes the standard `help()` functionality as found in the repl which this also applies to.
 2. As a note. I somewhat disagree with the error silencing implementation. If there's an error I'd either expect an exception to be thrown or `None` instead of `""`, `-1` etc. I'm a polyglot-programmer so I might be intermixing conventions. This could be explained better in docs as well as have code comments for contributors implementing custom scraper to maintain convention.

**Comment by jayaddison** [comment]
Re: exception silencing: I tend to be in the same boat as you about 'raising errors by default', but it's also true that web scraping can be very unpredictable.  Perhaps there is a difference between "the scraper did not find what it was looking for as instructed by the code" and "an error occurred while running the scraper code for this field" - maybe we could provide sensible defaults in the former case?  And/or a way for the developer to know that a default was returned?

I'm still thinking about abstract classes and what the implications of using those might be; I haven't used them recently and forget some of the details about how they behave.  Mainly it feels worth being careful not to introduce changes that make reading or writing scrapers more challenging than it needs to be.

**Comment by bfcarpio** [comment]
> Mainly it feels worth being careful not to introduce changes that make reading or writing scrapers more challenging than it needs to be.

I agree that this something to keep in mind. The generator script alone reduces the labour required for creating scrapers. I think the biggest benefit in this regard of abstract classes is that we could force consistency. Tests will fail if not all the functions are overridden and good code comments in the templates can guide users. For example, if there are no ratings there could be a comment that says "If no ratings are available just `return None` for this function". Mind you, we don't have to require all the parameters be overridden. We could say that title, ingredients, and instructions are mandatory, but everything else is optional and will return `None` if not defined (as well as potentially other reasons etc).

**Comment by bfcarpio** [comment]
I made #318 for discussion on returning conventions.

Also, I updated the initial post to mention that doc hover in IDEs usually uses the same output of `help()` in REPL. When I did manual testing, I checked both sources.

**Comment by bfcarpio** [comment]
Closing this since it's incompatible with the metaclass structure we ended up going with. I have an idea on how I can get all the docs to show up for the function, but I need to test further if it actually shows up when developing or only exists during runtime. This is true for dynamic typing for the defaulting values as well.

A touch burned out by $work so I haven't contributed much recently. Hoping to get back into it soon. Thanks guys!
